services/search/generate_docs_emebddings.py

```diff
@@ -0,0 +1,125 @@
+import os
+import time
+import glob
+import argparse
+from dotenv import load_dotenv
+from langchain_text_splitters import RecursiveCharacterTextSplitter, MarkdownHeaderTextSplitter
+from pymilvus import FieldSchema, CollectionSchema, DataType, utility, Collection, connections, model
+
+load_dotenv()
+
+def read_md_files(directory):
+    md_files = glob.glob(f"{directory}/**/*.md", recursive=True)
+    docs = []
+    for file in md_files:
+        with open(file, "r", encoding="utf-8") as f:
+            docs.append((file, f.read()))  # Returning a tuple with file name and content
+    return docs
+
+def split_md_by_sections(file_name, content):
+    headers_to_split_on = [
```

Comment:
The problem here is that if we just split the documents based on headings it would result in very uneven chunks (some greater than 2700). Therefore, using a text splitter after chunking based on headers seems necessary. 

The character splitter provides a parameter `chunk_overlap` which helps us to maintain context between the chunks.

This helps in maintaining proportionate chunk size and also maintains context between the chunks. Also, only sections with greater size are split with some overlap and we can change the value to see what works for us.
```
    text_splitter = RecursiveCharacterTextSplitter(chunk_size=512, chunk_overlap=128, length_function=len, 
    is_separator_regex=False)
    splits = text_splitter.split_documents(md_header_splits)
```

Reply:
Great, thank you for flagging this! Please do raise issues as you find them and bring out discussion points in prominent places.

I am shutting down for the weekend but a couple of thoughts:

- Is uneven chunk size really a problem? Is it just a problem of input token size?
- Can we add metadata to the documentation to aid chunking?
- Would it help to re-write the docs to have more even sections? I have no problem re-writing docs to aid ML (so long as it doesn't hinder human learning, of course)

If we need to experiment with the split parameters, I would encourage you to spend some time on this. Focus on the job writing guide. You should be able to look at the chunks and, as a human, feel like it's describing a meaningful unit of knowledge. A fixed set of concepts. If you look a chunk at random - can you tell what it's talking about?

Are there any other pre or processing steps that could help here? You may want to research this

Reply:
- Yes, uneven chunking affects the search results and eventually the input token size.
- We can add metadata, but the overlap factor already adds context between the chunks.
- Yes, we can do that as well, but in future when we plan to embed all of the docs, this won't be a good approach, I guess. 
About the pre processing steps, I think it will be highly effective to deploy a fine tunned model trained on OpenFn datasets and call that model. Also, once this is implemented, we can use feedback loops to improve the results.